File: csharp/dnd-character/DndCharacter.cs
using System;
using System.Linq;
using System.Collections.Generic;

public class DndCharacter
{
    public DndCharacter(int strength, int dexterity, int constitution, int intelligence, int wisdom, int charisma)
    {
        Strength = strength;
        Dexterity = dexterity;
        Constitution = constitution;
        Intelligence = intelligence;
        Wisdom = wisdom;
        Charisma = charisma;
        Hitpoints = 10 + Modifier(constitution);
    }

    public int Strength { get; }
    public int Dexterity { get; }
    public int Constitution { get; }
    public int Intelligence { get; }
    public int Wisdom { get; }
    public int Charisma { get; }
    public int Hitpoints { get; }

    public static int Modifier(int score)
    {
        return (int)Math.Floor((double)(score - 10) / 2);
    }

    public static int Ability() 
    {
        var rolls = new List<int>();
        var random = new Random();
        for(int i = 0; i < 4; i++)
        {
            rolls.Add(random.Next(1, 7));
        }
        rolls.Remove(rolls.Min());
        return rolls.Sum();
    }

    public static DndCharacter Generate()
    {
        return new DndCharacter(Ability(), Ability(), Ability(), Ability(), Ability(), Ability());
    }
}

File: javascript/dnd-character/dnd-character.js
export const abilityModifier = (score) => {
  if(score > 18){
    throw new Error('Ability scores can be at most 18');
  }else if(score < 3){
    throw new Error('Ability scores must be at least 3');
  }
  return Math.floor((score - 10) / 2);
};

export class Character {
  //rolls dice 4 times, returns sum of 3 highest rolls
  static rollAbility() {
    let rolls = [];
    for(let i = 0; i < 4; i++){
      rolls.push(rollDice(1, 6));
    }
    return rolls.sort().slice(1).reduce((x, y) => x + y, 0);
  }
  
  constructor(){
    this.strength = Character.rollAbility();
    this.dexterity = Character.rollAbility();
    this.constitution = Character.rollAbility();
    this.intelligence = Character.rollAbility();
    this.wisdom = Character.rollAbility();
    this.charisma = Character.rollAbility();
    this.hitpoints = 10 + abilityModifier(this.constitution);
  }
}

const rollDice = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1)) + min;
}